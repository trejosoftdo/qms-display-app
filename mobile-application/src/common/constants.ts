export const AUTHORIZATION_PENDING_CODE = 'authorization_pending';

export const VALIDATION_ERROR_STATUS_CODE = 400;

export const DEVICE_NOT_CONNECTED_ERROR_MESSAGE = 'Device is not connected';

export const CONNECTION_EXPIRED_ERROR_MESSAGE = 'Connection expired!';

export const APP_PROTO = 'app://';

export const DEVICE_ID_KEY = 'SECURE_DEVICE_ID';

export const CONNECTION_DETAILS_KEY = 'CONNECTION_DETAILS';

export const WEB_PLATFORM = 'web';

export const BEARER_PORTION = 'Bearer';

export const EMPTY_VALUE = '';

export const DEFAULT_REGULAR_FONT_NAME = 'Montserrat-Regular';

export const DEFAULT_LIGHT_FONT_NAME = 'Montserrat-Light';

export const DEFAULT_MEDIUM_FONT_NAME = 'Montserrat-Medium';

export const NORMAL_FONT_WEIGHT = 'normal';

export const ONE_SECOND_MILISECONDS = 1000;

export const INTERVAL_TIME = 10 * ONE_SECOND_MILISECONDS;

export const SERVICES_PATH = '/services';

export const TICKET_DETAILS_PATH = '/ticket-details';

export const HOME_PATH = '/';

export const CATEGORIES_PATH = '/categories';

export const CONFIGURATION_PATH = '/configuration';

export const NO_OP = () => {};


export const LANGUAGES = {
  SPANISH: 'es',
  ENGLISH: 'en',
};

export const CONFIG_OPTIONS = {
  INFORMATION: 'information',
  CONNECTION: 'connection',
  LANGUAGE: 'language',
};
