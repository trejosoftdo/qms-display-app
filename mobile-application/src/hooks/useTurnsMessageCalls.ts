import { useEffect } from "react";
import { loadMultipleAudio } from "../common/services/service-turn";

/**
 * Calls turns via audio
 */
const useTurnsMessageCalls = (textItems: string[] | undefined): void => {
  useEffect(() => {
    if (textItems?.length) {
      loadMultipleAudio(textItems)
        .catch(console.error);
    }
  }, [textItems?.join('.')]);
};

export default useTurnsMessageCalls;
