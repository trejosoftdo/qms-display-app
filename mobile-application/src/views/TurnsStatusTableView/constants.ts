export const BEING_ATTENDED_STATUS_CODE = "BEING_ATTENDED";

export const TO_BE_ATTENDED_STATUS_CODE = "TO_BE_ATTENDED";