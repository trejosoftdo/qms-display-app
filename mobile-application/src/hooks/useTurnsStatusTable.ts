import { useProgress } from "../common/hooks";
import { TurnsStatusTableData, Progress } from "../common/models";
import { loadTurnsStatusTable } from "../common/services/service-turn";

/**
 * Hook to get the turns status table data
 * @returns Progress<TurnsStatusTableData>
 */
const useTurnsStatusTable = (): Progress<TurnsStatusTableData> => useProgress<TurnsStatusTableData>(
  loadTurnsStatusTable(),
  null,
  () => loadTurnsStatusTable(),
);

export default useTurnsStatusTable;
