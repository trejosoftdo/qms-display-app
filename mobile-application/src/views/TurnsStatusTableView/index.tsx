import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { View, StyleSheet } from 'react-native';
import { DataTable } from 'react-native-paper';
import { AppView, ConditionalContainer } from '../../common/components';
import useTurnsStatusTable from '../../hooks/useTurnsStatusTable';
import useTurnsMessageCalls from '../../hooks/useTurnsMessageCalls';
import {
  TRANSLATION_NEXT_TURNS_KEY,
  TRANSLATION_TURN_KEY,
  TRANSLATION_POSITION_KEY,
  TRANSLATION_WAIT_MESSAGE_KEY,
  TRANSLATION_TO_KEY,
} from '../../common/translations/translation-keys';
import { BEING_ATTENDED_STATUS_CODE, TO_BE_ATTENDED_STATUS_CODE } from './constants';

/**
 * TurnsStatusTableViewProps defines the props for the Component.
 */
interface TurnsStatusTableViewProps { }

/**
 * A component that represents the view to display the turns status table
 *
 * @param {TurnsStatusTableViewProps} props - The props for the component.
 */
const TurnsStatusTableView: React.FC<TurnsStatusTableViewProps> = (props: TurnsStatusTableViewProps) => {
  const {
    loading,
    data,
    error,
  } = useTurnsStatusTable();
  const { t } = useTranslation();
  const messageKey = loading ? TRANSLATION_WAIT_MESSAGE_KEY : "";
  const beingAttendedItems = data?.items?.filter(item => item.statusCode === BEING_ATTENDED_STATUS_CODE);
  const toBeAttendedItems = data?.items?.filter(item => item.statusCode === TO_BE_ATTENDED_STATUS_CODE);
  const textItems = toBeAttendedItems?.map(item => `${item.ticketNumber} ${t(TRANSLATION_TO_KEY)} ${item.queueName}`);
  
  if (textItems?.length) {
    textItems.unshift(t(TRANSLATION_NEXT_TURNS_KEY));
  }

  useTurnsMessageCalls(textItems);

  return (
    <AppView
      loading={loading}
      error={error}
    >
      <ConditionalContainer display={!loading && !!data}>
        <View style={styles.content}>
          <DataTable style={styles.beingAttended}>
            <DataTable.Header style={styles.header}>
              <DataTable.Title textStyle={styles.title}>
                {t(TRANSLATION_TURN_KEY)}
              </DataTable.Title>
              <DataTable.Title textStyle={styles.title}>
                {t(TRANSLATION_POSITION_KEY)}
              </DataTable.Title>
            </DataTable.Header>

            {beingAttendedItems?.map((item) => (
              <DataTable.Row key={item.ticketNumber}>
                <DataTable.Cell textStyle={styles.cell}>{item.ticketNumber}</DataTable.Cell>
                <DataTable.Cell textStyle={styles.cell}>{item.queueName}</DataTable.Cell>
              </DataTable.Row>
            ))}
          </DataTable>

          <DataTable style={styles.toBeAttended}>
            <DataTable.Header style={styles.header}>
              <DataTable.Title textStyle={styles.title}>
                {t(TRANSLATION_NEXT_TURNS_KEY)}
              </DataTable.Title>
            </DataTable.Header>

            {toBeAttendedItems?.map((item) => (
              <DataTable.Row key={item.ticketNumber}>
                <DataTable.Cell textStyle={styles.cell}>
                  {item.ticketNumber} {t(TRANSLATION_TO_KEY)} {item.queueName}
                </DataTable.Cell>
              </DataTable.Row>
            ))}
          </DataTable>
        </View>
      </ConditionalContainer>
    </AppView>
  );
};

const styles = StyleSheet.create({
  header: {
    backgroundColor: '#b4c5e4',
    textTransform: 'uppercase',
  },
  title: {
    fontSize: 20,
    fontWeight: "600",
    color: '#000',
  },
  cell: {
    fontSize: 18,
    fontWeight: "600",
    color: '#000',
  },
  content: {
    width: '100%',
    display: 'flex',
    flex: 1,
    flexDirection: 'row',
    alignItems: 'flex-start',
    justifyContent: 'space-between',
    paddingTop: 50
  },
  beingAttended: {
    width: '60%',
    backgroundColor: '#fff',
  },
  toBeAttended: {
    width: '30%',
    backgroundColor: '#fff',
  },
});

export default TurnsStatusTableView;