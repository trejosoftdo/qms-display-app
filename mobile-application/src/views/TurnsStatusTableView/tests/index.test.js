import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import useTurnsStatusTable from '../../../hooks/useTurnsStatusTable';
import useTurnsMessageCalls from '../../../hooks/useTurnsMessageCalls';
import TurnsStatusTableView from '../';
import { BEING_ATTENDED_STATUS_CODE, TO_BE_ATTENDED_STATUS_CODE } from '../constants';


jest.mock('expo-router');
jest.mock('../../../common/helpers');
jest.mock('../../../hooks/useTurnsStatusTable');
jest.mock('../../../hooks/useTurnsMessageCalls');


describe('TurnsStatusTableView component', () => {
  const mockParams = {
    customerName: 'customer-name',
  };
  const mockData = {
    items: [
      {
        ticketNumber: 'mock-first-ticket-number',
        queueName: 'mock-first-queue-name',
        statusCode: BEING_ATTENDED_STATUS_CODE,
      },
      {
        ticketNumber: 'mock-second-ticket-number',
        queueName: 'mock-second-queue-name',
        statusCode: TO_BE_ATTENDED_STATUS_CODE,
      },
    ],
  };

  afterEach(() => {
    jest.clearAllMocks();
  })

  it('renders the component as expected when loading', () => {
    useTurnsStatusTable.mockReturnValue({ loading: true });
    expect(renderShallow(<TurnsStatusTableView />)).toMatchSnapshot();
    expect(useTurnsStatusTable).toHaveBeenCalledTimes(1);
    expect(useTurnsStatusTable).toHaveBeenCalledWith();
    expect(useTurnsMessageCalls).toHaveBeenCalledTimes(1);
    expect(useTurnsMessageCalls).toHaveBeenCalledWith(undefined);
  });

  it('renders the component as expected when data has been loaded', () => {
    useTurnsStatusTable.mockReturnValue({
      loading: false,
      data: mockData,
    });
    expect(renderShallow(<TurnsStatusTableView />)).toMatchSnapshot();
    expect(useTurnsStatusTable).toHaveBeenCalledTimes(1);
    expect(useTurnsStatusTable).toHaveBeenCalledWith();
    expect(useTurnsMessageCalls).toHaveBeenCalledTimes(1);
    expect(useTurnsMessageCalls).toHaveBeenCalledWith([
      "Translated[translation:nextTurns]({})",
      "mock-second-ticket-number Translated[translation:to]({}) mock-second-queue-name"
    ]);
  });

  it('renders the component as expected when it has errors', () => {
    useTurnsStatusTable.mockReturnValue({
      loading: false,
      error: new Error('Unexpected error'),
    });
    expect(renderShallow(<TurnsStatusTableView />)).toMatchSnapshot();
    expect(useTurnsStatusTable).toHaveBeenCalledTimes(1);
    expect(useTurnsStatusTable).toHaveBeenCalledWith();
    expect(useTurnsMessageCalls).toHaveBeenCalledTimes(1);
    expect(useTurnsMessageCalls).toHaveBeenCalledWith(undefined);
  });
});
