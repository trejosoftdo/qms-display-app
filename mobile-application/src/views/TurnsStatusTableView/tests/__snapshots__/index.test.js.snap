// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TurnsStatusTableView component renders the component as expected when data has been loaded 1`] = `
<AppView
  loading={false}
>
  <ConditionalContainer
    display={true}
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "display": "flex",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingTop": 50,
          "width": "100%",
        }
      }
    >
      <DataTable
        style={
          {
            "backgroundColor": "#fff",
            "width": "60%",
          }
        }
      >
        <DataTable.Header
          style={
            {
              "backgroundColor": "#b4c5e4",
              "textTransform": "uppercase",
            }
          }
        >
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:turn]({})
          </DataTable.Title>
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:turn]({})
          </DataTable.Title>
        </DataTable.Header>
      </DataTable>
      <DataTable
        style={
          {
            "backgroundColor": "#fff",
            "width": "30%",
          }
        }
      >
        <DataTable.Header
          style={
            {
              "backgroundColor": "#b4c5e4",
              "textTransform": "uppercase",
            }
          }
        >
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:nextTurns]({})
          </DataTable.Title>
        </DataTable.Header>
      </DataTable>
    </View>
  </ConditionalContainer>
</AppView>
`;

exports[`TurnsStatusTableView component renders the component as expected when it has errors 1`] = `
<AppView
  error={[Error: Unexpected error]}
  loading={false}
>
  <ConditionalContainer
    display={false}
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "display": "flex",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingTop": 50,
          "width": "100%",
        }
      }
    >
      <DataTable
        style={
          {
            "backgroundColor": "#fff",
            "width": "60%",
          }
        }
      >
        <DataTable.Header
          style={
            {
              "backgroundColor": "#b4c5e4",
              "textTransform": "uppercase",
            }
          }
        >
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:turn]({})
          </DataTable.Title>
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:turn]({})
          </DataTable.Title>
        </DataTable.Header>
      </DataTable>
      <DataTable
        style={
          {
            "backgroundColor": "#fff",
            "width": "30%",
          }
        }
      >
        <DataTable.Header
          style={
            {
              "backgroundColor": "#b4c5e4",
              "textTransform": "uppercase",
            }
          }
        >
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:nextTurns]({})
          </DataTable.Title>
        </DataTable.Header>
      </DataTable>
    </View>
  </ConditionalContainer>
</AppView>
`;

exports[`TurnsStatusTableView component renders the component as expected when loading 1`] = `
<AppView
  loading={true}
>
  <ConditionalContainer
    display={false}
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "display": "flex",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingTop": 50,
          "width": "100%",
        }
      }
    >
      <DataTable
        style={
          {
            "backgroundColor": "#fff",
            "width": "60%",
          }
        }
      >
        <DataTable.Header
          style={
            {
              "backgroundColor": "#b4c5e4",
              "textTransform": "uppercase",
            }
          }
        >
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:turn]({})
          </DataTable.Title>
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:turn]({})
          </DataTable.Title>
        </DataTable.Header>
      </DataTable>
      <DataTable
        style={
          {
            "backgroundColor": "#fff",
            "width": "30%",
          }
        }
      >
        <DataTable.Header
          style={
            {
              "backgroundColor": "#b4c5e4",
              "textTransform": "uppercase",
            }
          }
        >
          <DataTable.Title
            textStyle={
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            Translated[translation:nextTurns]({})
          </DataTable.Title>
        </DataTable.Header>
      </DataTable>
    </View>
  </ConditionalContainer>
</AppView>
`;
